File: src/main/java/com/springdebugger/tap/BuildOutputAnalyzer.java
package com.springdebugger.tap;

import com.springdebugger.engine.DiagnosisPipeline;
import com.springdebugger.extractor.LogExtractor;
import com.springdebugger.model.DiagnosisCard;
import com.springdebugger.model.Phase;
import com.springdebugger.model.RawSignal;
import com.springdebugger.rule.RuleCatalog;

import java.util.Optional;

/**
 * The pure, IDE-free core shared by both build taps: it decides whether a chunk of
 * build output looks Spring-specific, and if so runs it through the diagnosis pipeline
 * at COMPILE phase. Keeping this free of any IntelliJ build API means it is fully
 * unit-testable with canned build output, which is the part of M6 we can actually verify
 * automatically (the live tap firing must be checked in a running IDE).
 */
public final class BuildOutputAnalyzer {

    private final LogExtractor extractor = new LogExtractor();
    private final DiagnosisPipeline pipeline;

    public BuildOutputAnalyzer(RuleCatalog catalog) {
        this.pipeline = new DiagnosisPipeline(catalog);
    }

    public Optional<DiagnosisCard> analyze(String buildOutput) {
        if (buildOutput == null || buildOutput.isBlank()) return Optional.empty();
        if (!isSpringRelated(buildOutput)) return Optional.empty();
        RawSignal signal = extractor.extract(buildOutput, Phase.COMPILE);
        return pipeline.run(signal);
    }

    /**
     * Generic Java compile errors are IntelliJ's job, not ours. This gate keeps the build
     * taps quiet unless the output carries a marker for one of the Spring-specific compile
     * rules (sections 6.4, 10.x, 13.x). It is deliberately broad on the marker side and
     * lets the rule catalog make the precise call.
     */
    static boolean isSpringRelated(String text) {
        return containsAny(text,
                "WebSecurityConfigurerAdapter",
                "spring-boot-configuration-processor",
                "lombok", "Lombok",
                "mapstruct", "MapStruct", "MapperImpl",
                "Unmapped target property",
                "Can't map property",
                "No implementation type is registered for return type",
                "Could not generate implementation",
                "UnsupportedClassVersionError",
                "class file has wrong version",
                "no main manifest attribute");
    }

    private static boolean containsAny(String text, String... needles) {
        for (String n : needles) {
            if (text.contains(n)) return true;
        }
        return false;
    }
}

File: vscode-extension/src/engine/build-output-analyzer.ts
// Port of com.springdebugger.tap.BuildOutputAnalyzer. Decides whether a chunk of build output is
// Spring-specific (generic Java compile errors are the IDE's job) and, if so, classifies it at
// COMPILE phase. The offline pipeline here is rule-engine-only, matching the headless Java path.
import { extract } from './log-extractor';
import { DiagnosisCard } from './models';
import { RuleBasedClassifier } from './rule-based-classifier';
import { RuleCatalog } from './rule-catalog';

const SPRING_MARKERS = [
  'WebSecurityConfigurerAdapter',
  'spring-boot-configuration-processor',
  'lombok',
  'Lombok',
  'mapstruct',
  'MapStruct',
  'MapperImpl',
  'Unmapped target property',
  "Can't map property",
  'No implementation type is registered for return type',
  'Could not generate implementation',
  'UnsupportedClassVersionError',
  'class file has wrong version',
  'no main manifest attribute',
];

export class BuildOutputAnalyzer {
  private readonly classifier: RuleBasedClassifier;

  constructor(catalog: RuleCatalog) {
    this.classifier = new RuleBasedClassifier(catalog);
  }

  analyze(buildOutput: string | null): DiagnosisCard | null {
    if (buildOutput === null || buildOutput.trim() === '') return null;
    if (!isSpringRelated(buildOutput)) return null;
    const signal = extract(buildOutput, 'COMPILE');
    return this.classifier.classify(signal);
  }
}

export function isSpringRelated(text: string): boolean {
  return SPRING_MARKERS.some((m) => text.includes(m));
}
